### DISCUSSION TITLE
experimental option "heroes: random skills on level up"

### DISCUSSION THREAD

**Discussion by viy2** [ihhub/fheroes2#4917] [Show and tell]
This is discussion for the pull request  https://github.com/ihhub/fheroes2/pull/4911
experimental option "heroes: random skills on level up"
It means player can change its hero's choice of skills on level up using the magic powers of save & load+random.
In fact, what I propose as an experimental option was the default behavior of fheroes2 before the 0.9.1.

This was changed in commit 9943758cf249bb8271148719aadbe17a4a22f3d0
(by vincent-grosbois at Wed Feb 17 2021 Deterministic bonus for hero level-up (#2630)  close #1973 )
the reason is to faithfully recreate the original hmm2.
However, I do like the old behavior and want to keep it at least as an option.

**Comment by LeHerosInconnu** [comment]
Hello @viy2,

From what I understand you are looking for the "perfect" hero. :)

And your proposal is to reload a savegame before the hero levels up to be able to do it (with the choice of random seconday skills each time).
I think this way would be a bit constraining and annoying for the player.

I see several possibilities for this concern:
- Make all secondary skills more interesting (https://github.com/ihhub/fheroes2/discussions/2990).
- Allow the hero to refuse the secondary skills proposed (the hero goes up in level but doesn't learn a secondary skill; a button is added to refuse the two seconday skills proposed).
- Propose new secondary skills if the two proposed secondary skills are not suitable (a button is added to propose two different secondary skills).
This last proposal corresponds most to the pursuit of the "perfect" hero but without the problem of having to reload a save and possibly redo a combat if the level up occurs at the end of a combat.

I think that it would take away some spice from the game if the hero could be exactly as you want. :)
But I still prefer the addition of a button to reroll the proposed secondary skills as a possibility rather than forcing the user to reload a save and possibly redo a combat, and it may be several times to achieve its ends.

**Comment by viy2** [comment]
> This last proposal corresponds most to the pursuit of the "perfect" hero but without the problem of having to reload a save and possibly redo a combat if the level up occurs at the end of a combat.

well, this would be even better.

> I think that it would take away some spice from the game if the hero could be exactly as you want. :)
That's why I think that this mode should be an option. The way different people get fun is different.

But I still prefer the addition of a button to reroll the proposed secondary skills as a possibility rather than forcing the user to reload a save and possibly redo a combat, and it may be several times to achieve its ends.

Indeed, this would be much better. Unfortunately, I did not dive deep into the code yet to be able to change that dialog to add the re-roll button.

**Reply by Branikolog** [reply]
Hi, @viy2!

I agree with @LeHerosInconnu here:

>I think that it would take away some spice from the game if the hero could be exactly as you want. :)

HoMM games combine strictly predictable mechanics as well as random ones. If we allow player to reroll secondary skills it will result in a situation, when ALL heroes will be the same and this will completely ruin game experience, making gameplay boring.
We shouldn't allow to reroll secondary skills (except, probably, the cases, it will be implemented via special mechanic, object or something  that will be working as an event, that couldn't be abused in game). 
The game has always has a certain pool of skills and artifacts, where some of them are good, and some are not, so player should make a choice (Sometimes - not easy one, for instance, between Eagle Eye and Navigation on a map without water). That is a key to "replayness" of the game. Without that - each game will look almost the same to previous one.

**Comment by viy2** [comment]
> this will completely ruin game experience, making gameplay boring

It depends on people. I agree that you will feel that way, and there are a lot of people who feel the same. But because of that I do propose it as an **non-default** option for the minority of people like me who do not like their OSD scratched and have their fun with re-rolling.

My original patch was for people motivated enough to do re-rolling using a lot of load and save.
But your argument interestingly applies against the ready "re-roll" button, that was proposed by LeHerosInconnu above.
May be for some people a ready "re-roll" button will too easy and will rob them their accomplishment of the long and hard cheat battle for the proper re-roll with lots of save & load ?

**Reply by Branikolog** [reply]
I understand, why players like you want a reroll for secondary skills, but I'm afraid, this can cause a significant interest reduction for the game. The same, as happens while cheat abusing.

**Reply by viy2** [reply]
> but I'm afraid, this can cause a significant interest reduction for the game. 

As default behavior, yes. But for a hidden experimental option? Not at all.
Those who have pleasure of hard choices just will not play with strange experimental options.

> The same, as happens while cheat abusing.

It all depends of proportion of cheats. Save and load are cheats too ;)
Well, anyway games are not healthy and cause addiction. We like them for fun, not for other reasons.
And what is fun is different for different people.

I do lost a lot of fun with 0.9.1 changes and it motivated me enough to wrote the proposed pull request.

**Reply by Branikolog** [reply]
It's only my personal opinion. If other players and contributors will vote for this option, we will add it into the engine. But note, that we don't want this engine to have unlimited number of options and adjustments, like Heroes3 WOG has. It will definitely scare the majority of players. Only the most valuable and requested should be in game. At least for early release.
I hope lots of mods, dueling versions and other expansions will appear later, based on our engine.

**Reply by viy2** [reply]
> At least for early release.

Yes, I know that I should wait for 1.0.
Ihar Hubchyk (@ihhub) said "we strictly don't accept any new experimental options till the game is ready for 1.0 release."
